service-worker/worker/src/plugins/dynamic/linked.ts

```diff
@@ -0,0 +1,222 @@
+   */
+  constructor(public compare: CompareFn<T>) {}
+
+  /**
+   * Head of the list, which is null if there are no elements.
+   */
+  head: Node<T>|null = null;
+
+  /**
+   * Tail of the list, which is null if there are no elements.
+   */
+  tail: Node<T>|null = null;
+
+  /**
+   * Tracks the current length of the list.
+   */
+  length: number = 0;
+
+  /**
+   * Insert a new element in a position determined by the
+   * comparison function.
+   *
+   * This is O(n).
+   */
+  insert(value: T): void {
+    // No matter what, inserting will increase the length by one.
+    this.length++;
+
+    // Special case insertion into an empty list.
+    if (this.head === null) {
+      // The head and tail will be the same node, which will have
+      // no siblings.
+      this.head = this.tail = {value, next: null, prev: null};
+      return;
+    }
+
+    // Scan through the list until `curr` becomes the node after the
+    // point where `value` is to be inserted.
+    let curr = this.head;
+    while (curr !== null) {
+      const cmp = this.compare(value, curr.value);
+      if (cmp <= 0) {
+        // `value` must be inserted before `curr`.
+        // At this point though, `curr` could still be at the head
+        // of the list, which means that `value` should be the new
+        // head.
+        if (curr.prev === null) {
+          // `curr` is indeed the current head. Construct a new node
+          // for `value` which has its next sibling as the current
+          // head, and assign it to the current head's previous
+          // sibling reference and also to the list head reference
+          // in one assignment statement.
+          // In other words, go from:
+          //   curr <-> ...
+          // to:
+          //   N(value) <-> curr <-> ...
+          this.head = this.head.prev = {value, next: this.head, prev: null};
+        } else {
+          // `curr` is in the middle of the list, so split `curr` and
+          // `curr.prev` and insert the new node in between.
+          // In other words, go from:
+          //   ... <-> prev <-> curr <-> ...
+          // to:
+          //   ... <-> prev <-> N(value) <-> curr <-> ...
+          curr.prev = curr.prev.next = {value, next: curr, prev: curr.prev};
+        }
+        // New value has been inserted successfully.
+        return;
+      }
+      
+      // Keep iterating through the list until the end is reached or
+      // the value is inserted.
+      curr = curr.next;
+    }
+
+    // Off the end of the list. This means the value belongs after
+    // the list tail.
+    // Construct the new node to go after the tail, and assign it in
+    // one statement.
+    // In other words, go from:
+    //   ... <-> tail
+    // to:
+    //   ... <-> tail <-> N(value)
+    this.tail = this.tail.next = {value, next: null, prev: this.tail};
+  }
+
+  /**
+   * Remove a value from the list.
+   * 
+   * This is O(n).
+   */
+  remove(value: T): void {
```

Comment:
It's not a concern, since a map ensures uniqueness of the URLs which are keyed in here.